sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py

```diff
@@ -0,0 +1,555 @@
+            full_feature_names=full_feature_names,
+            on_demand_feature_views=OnDemandFeatureView.get_requested_odfvs(
+                feature_refs, project, registry
+            ),
+            metadata=RetrievalMetadata(
+                features=feature_refs,
+                keys=list(set(entity_schema.keys()) - {event_timestamp_col}),
+                min_event_timestamp=entity_df_event_timestamp_range[0],
+                max_event_timestamp=entity_df_event_timestamp_range[1],
+            ),
+        )
+
+    @staticmethod
+    def pull_all_from_table_or_query(
+        config: RepoConfig,
+        data_source: DataSource,
+        join_key_columns: List[str],
+        feature_name_columns: List[str],
+        event_timestamp_column: str,
+        start_date: datetime,
+        end_date: datetime,
+    ) -> RetrievalJob:
+        """
+        Note that join_key_columns, feature_name_columns, event_timestamp_column, and created_timestamp_column
+        have all already been mapped to column names of the source table and those column names are the values passed
+        into this function.
+        """
+        assert isinstance(data_source, SparkSource)
+        warnings.warn(
+            "The spark offline store is an experimental feature in alpha development. "
+            "This API is unstable and it could and most probably will be changed in the future.",
+            RuntimeWarning,
+        )
+        from_expression = data_source.get_table_query_string()
+
+        field_string = (
+            '"'
+            + '", "'.join(
+                join_key_columns + feature_name_columns + [event_timestamp_column]
+            )
+            + '"'
+        )
+        start_date = start_date.astimezone(tz=utc)
+        end_date = end_date.astimezone(tz=utc)
+
+        query = f"""
+            SELECT {field_string}
+            FROM {from_expression}
+            WHERE "{event_timestamp_column}" BETWEEN TIMESTAMP '{start_date}' AND TIMESTAMP '{end_date}'
+        """
+        spark_session = get_spark_session_or_start_new_with_repoconfig(
+            store_config=config.offline_store
+        )
+        return SparkRetrievalJob(
+            spark_session=spark_session, query=query, full_feature_names=False
+        )
+
+
+# TODO fix internal abstract methods _to_df_internal _to_arrow_internal
+class SparkRetrievalJob(RetrievalJob):
+    def __init__(
+        self,
+        spark_session: SparkSession,
+        query: str,
+        full_feature_names: bool,
+        on_demand_feature_views: Optional[List[OnDemandFeatureView]] = None,
+        metadata: Optional[RetrievalMetadata] = None,
+    ):
+        super().__init__()
+        self.spark_session = spark_session
+        self.query = query
+        self._full_feature_names = full_feature_names
+        self._on_demand_feature_views = on_demand_feature_views
+        self._metadata = metadata
+
+    @property
+    def full_feature_names(self) -> bool:
```

Comment:
we've started moving towards just public attributes, so maybe remove these and just do self.metadata = metadata in the constructor, etc?